### ISSUE TITLE
Investigate computing crate_hash by hashing metadata

### ISSUE THREAD

**Issue by cjgillot** [rust-lang/rust#94878]
The `crate_hash` query is used to detect is metadata has changed before attempting to decode more complex information from it.
Until now, this crate hash has been computed manually using HIR.
Computing it while encoding metadata itself may be more robust and exhaustive.

General instructions:
- add a `StableHasher` and a `StableHashingContext` in `rustc_metadata::rmeta::encoder::EncodeContext`;
- double all accesses to the embedded `opaque` encoder by a call to `hash_stable` with the same data;
- manually reimplement encoding and hashing for special datatypes (`CrateNum`, `DefId`, `LocalDefId`...);
- remove the implementation of the `crate_hash` query in `rustc_middle::hir::map`, and reimplement it from the just-encoded metadata;
- evaluate effect on performance.

Please reach out on zulip if more details are required.
cc @michaelwoerister

**Comment by michaelwoerister** [comment]
I think this is very promising. I would suggest to entirely remove the local version of the `crate_hash` query. That should make things easier to implement.

**Comment by bjorn3** [comment]
The crate metadata is missing the session configuration with which the crate has been compiled. This means that eg changing optimization level probably won't change crate metadata. It probably needs to change the crate hash though.

**Comment by michaelwoerister** [comment]
Ideally it wouldn't be necessary because the hash includes everything that is accessible to downstream crates, right?

**Comment by bjorn3** [comment]
The object files are accessible to downstream crates (they are linked into the final artifact after all), but they depend on both the crate metadata and the compilation session options.

**Comment by michaelwoerister** [comment]
Yes, but linking is done unconditionally, independent of the crate hash.

**Comment by bjorn3** [comment]
I think it isn't too far fetched that cargo or other build systems will use the crate hash in the far future instead of mtime or hashing the entire file.

**Comment by michaelwoerister** [comment]
I don't know. In my opinion the crate hash is an internal implementation detail of the incremental compilation system that's only there to optimize a common code path there. I wouldn't recommend any outside tooling to rely on it.

**Comment by bjorn3** [comment]
The SVH has introduced feb 2014 in https://github.com/rust-lang/rust/commit/ec57db083ff10fc4da0a2f25df5acf1d4398e560

> The goal of this hash is to identify when upstream crates have changed but
downstream crates have not been recompiled.

**Comment by lqd** [comment]
(I've begun looking at this already, but I've unfortunately realized that I won't have time to work on this soon enough. 

Since it's both short-to-try and has mentorship available from michael and camille, it should be a good opportunity for a new contributor, rather than waiting on me and taking this opportunity away from someone who's interested.

@rustbot release-assignment
)

**Comment by PrestonFrom** [comment]
I'm still pretty new, but I'd like to take a stab at this. If it would be more appropriate for something with more experience, though, I'm happy to be un-assigned.

@rustbot claim

**Comment by PrestonFrom** [comment]
I've gotten very busy all the sudden -- I'm going to unassign myself in case someone has bandwidth to grab this now.

**Comment by Enselic** [comment]
Triage: Are you still working on this @DesmondWillowbrook ?

**Comment by pierwill** [comment]
I might be interested in taking this on :)

**Comment by Daniel-B-Smith** [comment]
I have a (mostly) working implementation of this: https://github.com/rust-lang/rust/pull/154724, but the performance impact is not great. Unless I am doing something wrong (certainly possible), the metadata hash is much more expensive than the bits of the HIR hash that can be removed. The implementation of `rustc_middle::hir::map::crate_hash` itself seems to be relatively cheap because most of the work is in `LocalCrate.opt_hir_hash`. The latter is computed in `rustc_ast_lowering::compute_hir_hash`. That in turn seems to be relatively cheap because most of the work is done computing the DefPathHashes. AFAIU, those hashes are needed elsewhere, so we can only remove `crate_hash` and `compute_hir_hash`. 

If there is interest in moving to a metadata hash because it would be more robust in spite of the performance hit, I'm willing to push forward in cleaning up that PR and looking for any additional computations that I can trim. Otherwise, it probably makes sense to close this out as a reasonable idea but not worth the performance implications.

For the sake of completeness, I attempted to add conditionals on only computing the hash when `tcx.needs_crate_hash()` is true and/or only serializing the metadata when `tcx.needs_metadata()` was true, but neither meaningfully improved performance (on my laptop at least).

**Comment by Daniel-B-Smith** [comment]
There were more performance optimizations that I could do. The tl;dr right now is that the change either accepts a small/moderate cost to IncrUnchanged with a small/moderate improvement to IncrPatched or vice versa:

IncrUnchanged more expensive: https://perf.rust-lang.org/compare.html?start=f53b654a8882fd5fc036c4ca7a4ff41ce32497a6&end=d73a8336e43c291c192ed07d1d4c5f82597e3d86&stat=instructions:u

vice versa: https://perf.rust-lang.org/compare.html?start=1d72d7e8136faaebad3a85eeed432e6ea1b2ffab&end=ad26d1abc0cd7a7ea9b98cf477d0f887f415fd55&stat=instructions:u

the underlying issue is that (AFAICT) the HIR hash is used for dependency analysis before the metadata is computed. if you drop the HIR hash altogether, dependency analysis gets rerun even in some unchanged cases. if you keep the HIR hash, you're doing extra work in the changed case to calculate the metadata hash from scratch.

I'm going to poke at making the HIR hash cheaper to make the trade off less stark, but I think the change will still be a mixed bag.
